admin/autoupdater/src/uclient.c

```diff
@@ -0,0 +1,144 @@
+
+    1. Redistributions of source code must retain the above copyright notice,
+       this list of conditions and the following disclaimer.
+    2. Redistributions in binary form must reproduce the above copyright notice,
+       this list of conditions and the following disclaimer in the documentation
+       and/or other materials provided with the distribution.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+#include "uclient.h"
+
+#include <libubox/uloop.h>
+
+#include <limits.h>
+#include <stdio.h>
+
+
+#define TIMEOUT_MSEC 300000
+
+static const char *const user_agent = "Gluon Autoupdater (using libuclient)";
+
+struct uclient_data {
+	/* data that can be passed in by caller and used in custom callbacks */
+	void *custom;
+	/* data used by uclient callbacks */
+	int retries;
+	int err_code;
+};
+
+inline struct uclient_data *uc_data(struct uclient *cl) {
+	return (struct uclient_data *)cl->priv;
+}
+
+enum uclient_own_error_code {
+	UCLIENT_ERROR_REDIRECT_FAILED = 32,
+	UCLIENT_ERROR_TOO_MANY_REDIRECTS,
+	UCLIENT_ERROR_CONNECTION_RESET_PREMATURELY,
+	UCLIENT_ERROR_STATUS_CODE = 1024,
+};
+
+
+const char *uclient_get_errmsg(int code) {
+	static char http_code_errmsg[16];
+	if (code & UCLIENT_ERROR_STATUS_CODE) {
+		snprintf(http_code_errmsg, 16, "HTTP error %d",
+			code & (~UCLIENT_ERROR_STATUS_CODE));
+		return http_code_errmsg;
+	}
+	switch(code) {
+	case UCLIENT_ERROR_CONNECT:
+		return "Connection failed";
+	case UCLIENT_ERROR_TIMEDOUT:
+		return "Connection timed out";
+	case UCLIENT_ERROR_REDIRECT_FAILED:
+		return "Failed to redirect";
+	case UCLIENT_ERROR_TOO_MANY_REDIRECTS:
+		return "Too many redirects";
+	case UCLIENT_ERROR_CONNECTION_RESET_PREMATURELY:
+		return "Connection reset prematurely";
+	default:
+		return "Unknown error";
+	}
+}
+
+
+static void request_done(struct uclient *cl, int err_code) {
+	uc_data(cl)->err_code = err_code;
+	uclient_disconnect(cl);
+	uloop_end();
+}
+
+
+static void header_done_cb(struct uclient *cl) {
+	if (uc_data(cl)->retries < 10) {
+		int ret = uclient_http_redirect(cl);
+		if (ret < 0) {
+			request_done(cl, UCLIENT_ERROR_REDIRECT_FAILED);
+			return;
+		}
+		if (ret > 0) {
+			uc_data(cl)->retries++;
+			return;
+		}
+	}
+	uc_data(cl)->retries = 0;
```

Comment:
Since we always use the `get_url()` function, which creates a new uclient instance for every call, it never has. Otherwise the redirects would accumulate over multiple URLs in a single instance, which uclient supports IIRC. And basically, I more or less borrowed this code from ufetch, and didn't think it was useless at the time. I'll remove it.